### DISCUSSION TITLE
Styled Property in UserControl  is not working

### DISCUSSION THREAD

**Discussion by RaoHammas** [AvaloniaUI/Avalonia#17159] [Q&A]
**Hello Avalonia experts,** 

I've tried pretty much everything I could find on internet and failed to make this work. I'm simply trying to set `Color` and `Icon` properties in the `UserControl` when `Trend` property changes. `OnTrendChange` method is called whenever the Trend property changes, and Color and Icon are set but on UI nothing changes. What's wrong here?

Secondly, is there a better and simple way to achieve the same thing? I noticed, i can't implement `INotifyProeprtyChange` on a `UserControl` in Avalonia. it says, "it's already implemented". Which is different from WPF where we could use simple CLR properties and trigger change on UI using `INotifyPropertyChange` interface. Can't we simply do same in Avalonia?

In Avalonia, I have to use `Styled` Property even though I don't want user to set these two properties `Color` and `Icon` from XAML. They are supposed to set based on Trend property.

So, what am I missing here? Quick help is appreciated. Thanks!

```
public partial class OpenTickerTrendControl : UserControl
{
    public static readonly StyledProperty<string> IconProperty =
        AvaloniaProperty.Register<OpenTickerTrendControl, string>(nameof(Icon));

    public string Icon
    {
        get => GetValue(IconProperty);
        set => SetValue(IconProperty, value);
    }

    public static readonly StyledProperty<IImmutableSolidColorBrush> ColorProperty =
        AvaloniaProperty.Register<OpenTickerTrendControl, IImmutableSolidColorBrush>(nameof(Color));

    public IImmutableSolidColorBrush Color
    {
        get => GetValue(ColorProperty);
        set => SetValue(ColorProperty, value);
    }

    public static readonly StyledProperty<OpenTickerTrend> TrendProperty =
        AvaloniaProperty.Register<OpenTickerTrendControl, OpenTickerTrend>(nameof(Trend));

    public OpenTickerTrend Trend
    {
        get => GetValue(TrendProperty);
        set => SetValue(TrendProperty, value);
    }

    public OpenTickerTrendControl()
    {
        InitializeComponent();
        TrendProperty.Changed.AddClassHandler<OpenTickerTrendControl>(OnTrendChange);
    }


    private void OnTrendChange(OpenTickerTrendControl openTickerTrendControl, AvaloniaPropertyChangedEventArgs args)
    {
        switch (args.NewValue)
        {
            case OpenTickerTrend.Up:
                Icon = "△"; // Up arrow icon
                Color = Brushes.Green;
                break;
            case OpenTickerTrend.Down:
                Icon = "▽"; // Down arrow icon
                Color = Brushes.Red;
                break;
            default:
                Icon = "--"; // Dash icon
                Color = Brushes.Gray;
                break;
        }
    }
}
```  

**XAML:**
```
<UserControl
   .....
   .....
   .....
    d:DesignHeight="450"
    d:DesignWidth="800"
    x:DataType="controls:OpenTickerTrendControl"
    mc:Ignorable="d">
    <Grid>
        <Grid.Styles>
            <Style Selector="TextBlock.icon">
                <Setter Property="FontFamily" Value="Segoe MDL2 Assets" />
                <Setter Property="FontSize" Value="16" />
                <Setter Property="VerticalAlignment" Value="Center" />
                <Setter Property="Margin" Value="2" />
            </Style>
        </Grid.Styles>

        <TextBlock
            Grid.Column="0"
            Foreground="{Binding Color}"
            Text="{Binding Icon}"
            ToolTip.Tip="{Binding Icon}" />
    </Grid>
</UserControl>
```

**Usage:**

```
<controls:OpenTickerTrendControl
    Grid.Row="3"
    Grid.Column="2"
    Trend="{Binding TrendInfo.LastTrend, UpdateSourceTrigger=PropertyChanged}" />
```

**Comment by stevemonaco** [comment]
Your bindings in the `UserControl` appear to be incorrectly referencing the `DataContext` (ie. typically the ViewModel). Instead you want something like:

```xml
<UserControl x:Name="root">
    ...
        <TextBlock
            Grid.Column="0"
            Foreground="{Binding #root.Color}"
            Text="{Binding Icon}"
            ToolTip.Tip="{Binding #root.Icon}" />
```

Could also use `"{Binding $parent[UserControl].Color}"` I think.

**Reply by RaoHammas** [reply]
But `ViewModel` shouldn't be required. The code behind of XAML should be the default data context. Isn't it in Avalonia like WPF?

And thanks for help, `x:Name=''Root'` works. But, why does this work and no other way? even `$parent` doesn't work.
Lastly, I don't want user to access `Color` and `Icon` props in XAML like `Trend` is accessible. So, is there a way you know to not use Styled prop for these to props?

**Reply by stevemonaco** [reply]
This is the same as WPF. `DataContext` should be reserved for the ViewModel. If this were instead a `TemplatedControl` and you were in a `ControlTemplate`, then you could use `TemplateBinding` instead.

If Color/Icon need to be hidden, then don't expose them as properties. Instead, obtain the `TextBlock` control reference and assign the properties manually in the change handler. In a `UserControl`, you can still use `x:Name` on the `TextBlock` to get the control reference in code-behind.

If this were instead a `TemplatedControl` and you were in a `ControlTemplate`, then you could use `TemplateBinding` instead. You would need to find the control, typically done in `OnApplyTemplate`, and mark the control with `[TemplatePart]` attribute:

https://github.com/AvaloniaUI/Avalonia/blob/981f47f05035b700a3b723653d887f338dfbcc1b/src/Avalonia.Controls/SplitButton/SplitButton.cs#L234-L242

**Reply by RaoHammas** [reply]
"This is the same as WPF" ...? 
In WPF, you can set `DataContext = this;` in constructor of code-behind file and you are good to go. UI will identify the context and property changes will work without any issue. In Avalonia, it's not working.

but again, I'm not understanding why `x:Name=Root` works, but not `$parent` and direct binding of property working...

**Reply by stevemonaco** [reply]
> In WPF, you can set DataContext = this; in constructor of code-behind file and you are good to go.

Same in Avalonia. It's just a bad design idea, so I don't advise using that approach. However, you didn't make this assignment in the code shown. Therefore, you get an inherited `DataContext` which is typically the ViewModel and how it should be.

> but not $parent

Since the property is on `OpenTickerTrendControl` and not `UserControl`, you might need to use `$parent[c:OpenTickerTrendControl]` and define `xmlns:c` to be where the control is located.
